File: src/main/java/dark/leech/text/plugin/security/ZipSecurityValidator.java
package dark.leech.text.plugin.security;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Enumeration;
import java.util.zip.ZipEntry;
import java.util.zip.ZipFile;

/**
 * Security validator for ZIP file operations. Prevents ZIP Slip attacks and enforces size limits.
 */
public class ZipSecurityValidator {

    private static final long MAX_ZIP_SIZE_BYTES = 100 * 1024 * 1024; // 100MB
    private static final long MAX_UNCOMPRESSED_SIZE_BYTES = 500 * 1024 * 1024; // 500MB
    private static final int MAX_ENTRY_COUNT = 1000;

    /**
     * Validates a ZIP file for security threats.
     *
     * @param zipPath Path to ZIP file
     * @throws SecurityValidationException if validation fails
     */
    public static void validateZipFile(Path zipPath) throws SecurityValidationException {
        if (!Files.exists(zipPath)) {
            throw new SecurityValidationException("ZIP file not found: " + zipPath);
        }

        // Check ZIP file size
        try {
            long zipSize = Files.size(zipPath);
            if (zipSize > MAX_ZIP_SIZE_BYTES) {
                throw new SecurityValidationException(
                        String.format(
                                "ZIP file too large: %d MB (max: %d MB)",
                                zipSize / (1024 * 1024), MAX_ZIP_SIZE_BYTES / (1024 * 1024)));
            }
        } catch (IOException e) {
            throw new SecurityValidationException("Failed to check ZIP file size", e);
        }

        // Validate all entries
        try (ZipFile zipFile = new ZipFile(zipPath.toFile())) {
            validateZipEntries(zipFile);
        } catch (IOException e) {
            throw new SecurityValidationException("Failed to open ZIP file", e);
        }
    }

    /**
     * Validates all ZIP entries for path traversal and size limits. Uses proper path normalization
     * to prevent ZIP Slip attacks.
     *
     * @param zipFile Open ZipFile to validate
     * @throws SecurityValidationException if validation fails
     */
    public static void validateZipEntries(ZipFile zipFile) throws SecurityValidationException {
        int entryCount = 0;
        long totalUncompressedSize = 0;

        Enumeration<? extends ZipEntry> entries = zipFile.entries();
        while (entries.hasMoreElements()) {
            ZipEntry entry = entries.nextElement();
            entryCount++;

            if (entryCount > MAX_ENTRY_COUNT) {
                throw new SecurityValidationException(
                        "ZIP file contains too many entries: "
                                + entryCount
                                + " (max: "
                                + MAX_ENTRY_COUNT
                                + ")");
            }

            // Check uncompressed size
            long uncompressedSize = entry.getSize();
            if (uncompressedSize > 0) {
                totalUncompressedSize += uncompressedSize;
                if (totalUncompressedSize > MAX_UNCOMPRESSED_SIZE_BYTES) {
                    throw new SecurityValidationException(
                            String.format(
                                    "Total uncompressed size too large: %d MB (max: %d MB)",
                                    totalUncompressedSize / (1024 * 1024),
                                    MAX_UNCOMPRESSED_SIZE_BYTES / (1024 * 1024)));
                }
            }

            // Validate entry name with path normalization
            validateEntryName(entry.getName());
        }
    }

    /**
     * Validates a ZIP entry name to prevent path traversal attacks. Uses Path normalization to
     * detect ZIP Slip attempts.
     *
     * @param entryName The ZIP entry name
     * @throws SecurityValidationException if path traversal detected
     */
    public static void validateEntryName(String entryName) throws SecurityValidationException {
        if (entryName == null || entryName.isEmpty()) {
            throw new SecurityValidationException("ZIP entry name is empty");
        }

        // Normalize the path and check if it escapes the current directory
        Path normalizedPath;
        try {
            normalizedPath = Paths.get(entryName).normalize();
        } catch (Exception e) {
            throw new SecurityValidationException("Invalid ZIP entry name: " + entryName, e);
        }

        // Check if normalized path starts with ".." (escapes current directory)
        if (normalizedPath.startsWith("..")) {
            throw new SecurityValidationException(
                    "ZIP entry escapes directory (path traversal): " + entryName);
        }

        // Check if normalized path is absolute
        if (normalizedPath.isAbsolute()) {
            throw new SecurityValidationException("ZIP entry contains absolute path: " + entryName);
        }

        // Additional check: look for obvious traversal patterns before normalization
        if (entryName.contains("../") || entryName.contains("..\\")) {
            throw new SecurityValidationException(
                    "ZIP entry contains path traversal sequence: " + entryName);
        }

        // Check for drive letter (Windows absolute path)
        if (entryName.matches("^[A-Za-z]:.*")) {
            throw new SecurityValidationException(
                    "ZIP entry contains Windows absolute path: " + entryName);
        }

        // Check for UNC paths
        if (entryName.startsWith("\\\\") || entryName.startsWith("//")) {
            throw new SecurityValidationException("ZIP entry contains UNC path: " + entryName);
        }
    }

    /**
     * Resolves a ZIP entry name safely against a base directory. Ensures the resolved path stays
     * within the base directory.
     *
     * @param baseDir Base directory
     * @param entryName ZIP entry name
     * @return Safe resolved path
     * @throws SecurityValidationException if path escapes base directory
     */
    public static Path resolveSafePath(Path baseDir, String entryName)
            throws SecurityValidationException {
        // Validate entry name first
        validateEntryName(entryName);

        // Resolve against base directory
        Path resolved = baseDir.resolve(entryName).normalize();

        // Ensure resolved path is still within base directory
        if (!resolved.startsWith(baseDir.normalize())) {
            throw new SecurityValidationException(
                    "ZIP entry path escapes base directory: " + entryName);
        }

        return resolved;
    }
}

File: engine/src/commonMain/kotlin/dev/haipham22/leechtext/plugin/security/ZipSecurityValidator.kt
package dev.haipham22.leechtext.plugin.security

import dev.haipham22.leechtext.log.EngineLogger
import dev.haipham22.leechtext.util.ZipIo
import okio.FileSystem
import okio.Path

/**
 * Validator ZIP chống ZIP Slip + size limit (port từ ZipSecurityValidator.java; P5.2c:
 * java.util.zip.ZipFile → ZipIo seam, entry count/size giữ nguyên).
 */
object ZipSecurityValidator {
    private const val MAX_ZIP_SIZE_BYTES = 100L * 1024 * 1024 // 100MB
    private const val MAX_UNCOMPRESSED_SIZE_BYTES = 500L * 1024 * 1024 // 500MB
    private const val MAX_ENTRY_COUNT = 1000

    /** Validate ZIP file cho security threats. */
    @Throws(SecurityValidationException::class)
    @Suppress("ThrowsCount") // validation chain ở trust boundary — mỗi guard ném exception có message riêng cho UI
    fun validateZipFile(
        zipPath: Path,
        log: EngineLogger,
    ) {
        if (!FileSystem.SYSTEM.exists(zipPath)) {
            throw SecurityValidationException("ZIP file not found: $zipPath")
        }

        // Check kích thước file zip
        val zipSize =
            try {
                FileSystem.SYSTEM.metadata(zipPath).size ?: 0L
            } catch (e: Exception) {
                throw SecurityValidationException("Failed to check ZIP file size", e)
            }
        if (zipSize > MAX_ZIP_SIZE_BYTES) {
            throw SecurityValidationException(
                "ZIP file too large: ${zipSize / (1024 * 1024)} MB" +
                    " (max: ${MAX_ZIP_SIZE_BYTES / (1024 * 1024)} MB)",
            )
        }

        // Validate mọi entry
        try {
            validateZipEntries(zipPath, log)
        } catch (e: SecurityValidationException) {
            throw e
        } catch (e: Exception) {
            throw SecurityValidationException("Failed to open ZIP file", e)
        }
    }

    /**
     * Validate mọi ZIP entry cho path traversal + size limit. Path normalization chống ZIP
     * Slip.
     */
    @Throws(SecurityValidationException::class)
    @Suppress("ThrowsCount") // validation chain ở trust boundary — mỗi guard ném exception có message riêng cho UI
    fun validateZipEntries(
        zipPath: Path,
        log: EngineLogger,
    ) {
        var entryCount = 0
        var totalUncompressedSize = 0L

        val entries =
            try {
                ZipIo.entries(zipPath, log)
            } catch (e: Exception) {
                throw SecurityValidationException("Failed to read ZIP entries", e)
            }

        for (entry in entries) {
            entryCount++

            if (entryCount > MAX_ENTRY_COUNT) {
                throw SecurityValidationException("ZIP file contains too many entries: $entryCount (max: $MAX_ENTRY_COUNT)")
            }

            // Check uncompressed size
            if (entry.size > 0) {
                totalUncompressedSize += entry.size
                if (totalUncompressedSize > MAX_UNCOMPRESSED_SIZE_BYTES) {
                    throw SecurityValidationException(
                        "Total uncompressed size too large: " +
                            "${totalUncompressedSize / (1024 * 1024)} MB" +
                            " (max: ${MAX_UNCOMPRESSED_SIZE_BYTES / (1024 * 1024)} MB)",
                    )
                }
            }

            // Validate entry name với path normalization
            validateEntryName(entry.name)
        }
    }

    /** Validate ZIP entry name chống path traversal (ZIP Slip). */
    @Throws(SecurityValidationException::class)
    @Suppress("ThrowsCount") // validation chain ở trust boundary — mỗi guard ném exception có message riêng cho UI
    fun validateEntryName(entryName: String?) {
        if (entryName.isNullOrEmpty()) {
            throw SecurityValidationException("ZIP entry name is empty")
        }

        // Normalize path (bỏ ./ và resolve ../), check escape khỏi current directory
        val normalizedPath = entryName.replace("\\", "/")
        val segments = normalizedPath.split("/")
        var depth = 0
        for (seg in segments) {
            when (seg) {
                // segment rỗng (split "//") và "." — không đổi depth
                "", "." -> depth += 0

                ".." -> {
                    depth--
                    if (depth < 0) {
                        throw SecurityValidationException("ZIP entry escapes directory (path traversal): $entryName")
                    }
                }

                else -> depth++
            }
        }

        // Absolute path posix
        if (normalizedPath.startsWith("/")) {
            throw SecurityValidationException("ZIP entry contains absolute path: $entryName")
        }

        // Drive letter (Windows absolute path)
        if (entryName.matches(Regex("^[A-Za-z]:.*"))) {
            throw SecurityValidationException("ZIP entry contains Windows absolute path: $entryName")
        }

        // UNC path
        if (entryName.startsWith("\\\\") || entryName.startsWith("//")) {
            throw SecurityValidationException("ZIP entry contains UNC path: $entryName")
        }
    }

    /**
     * Resolve ZIP entry name an toàn với base dir — đảm bảo path nằm trong base dir.
     */
    @Throws(SecurityValidationException::class)
    fun resolveSafePath(
        baseDir: Path,
        entryName: String,
    ): Path {
        // Validate entry name trước
        validateEntryName(entryName)

        // Resolve với base directory (entryName đã validate không chứa ..)
        val resolved = baseDir / entryName

        // Đảm bảo vẫn nằm trong base directory
        if (!resolved.toString().startsWith(baseDir.toString())) {
            throw SecurityValidationException("ZIP entry path escapes base directory: $entryName")
        }

        return resolved
    }
}
